### PR TITLE
feat: Support for integer search space, lrq, cubic and lrqfa in vw-hyperopt

### PR DESCRIPTION
**Pull request by thmavri**
Three additions in utl/vw-hyperopt.py

Addition 1)
integer search space with ~ I, therefore options like the following work:  
--boosting=10..100
--bootstrap=2..50

Addition 2)
Two modes for search in lrq and lrqfa
i) same as -q for multiple choices of interactions, e.g. --lrq=ab2+bc3,ab2, --lrqfa=abc3,ab10
ii) finding optimal k, 
e.g. --lrq=ab2..10, --lrqfa=ab2..10 . 

It supports also for lrq to have multiple interactions and range together

e.g. --lrq=ab2..10+bc3..5 and --lrq==ab2..10+bc3..5,ab2..3+ac5..10

Addition 3)
Support for --cubic and --quadratic

### CODE REVIEW

File: utl/vw-hyperopt.py [hunk 1]
```diff
@@ -91,24 +92,30 @@ def _process_vw_argument(self, arg, value, algorithm):
...
+                else:
+                    raise ValueError("Cannot recognize distribution: %s" % (distr_part))
+
         else:
             possible_values = range_part.split(',')
-            if arg == '-q':
-                possible_values = [v.replace('+', ' -q ') for v in possible_values]
+            if arg == '--lrq' or arg == '--lrqfa' or arg == '-q' or arg == '--quadratic' or arg == '--cubic':
+                possible_values = [v.replace('+', ' {a} '.format(a=arg)) for v in possible_values]
+                print(possible_values)
```
[the remark this answers is not in the document]
Comment (thmavri) [reply utl/vw-hyperopt.py]: Removed the print. I was just debugging something.

File: utl/vw-hyperopt.py [hunk 2]
```diff
@@ -144,7 +151,11 @@ def string_to_pyll(self):
                     continue
                 if arg not in self.algorithm_metadata[algo]['prohibited_flags']:
                     distrib = self._process_vw_argument(arg, value, algo)
-                    self.space[algo][arg] = distrib
+                    key = arg
+                    if (arg=='--lrq' or arg=='--lrqfa') and '..' in value:
+                        ns = re.sub('[^a-zA-Z]+', '', value)
+                        key = arg +'='+ns
```
[the remark this answers is not in the document]
Comment (jackgerrits) [reply utl/vw-hyperopt.py:155]: the combination of a value and a range is a special case at least with respect to what exists currently. The value that goes with lrq or lrqfa is usually going to be quite small. In order to keep this special case constrained to the argument parsing section of the code (`_process_vw_argument`) you could simply create a choice distribution from enumerating each possible value. For example:

```python
ns = re.sub('[^a-zA-Z]+', '', value)
possible_values = [f"--lrq {ns}{v}" for v in range(vmin, vmax)]
distrib = hp.choice(hp_choice_name, possible_values)

**Comment by thmavri** [comment]
@jackgerrits what else do you propose I should do for this to be merged?

**Comment by jackgerrits** [comment]
@thmavri sorry for not getting back to you sooner. Looks good to me! I'll merge once the CI is done

**Review by jackgerrits — APPROVED** [review]

**Comment by jackgerrits** [comment]
Thanks!

**Pull request merged**